### PR TITLE
src: adds check_domain to handle size limits given by the specs

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
Hi guys. I'm opening this PR to check if this code (the content of this PR) is what is expected to solve this issue. I'll add tests for this soon. 

refs: https://github.com/ada-url/ada/issues/69

### CODE REVIEW

File: src/url.cpp [hunk 1]
```diff
@@ -469,6 +474,13 @@ namespace ada {
     if(checkers::is_ipv4(host.value())) {
       return parse_ipv4(host.value());
     }
+
+    // The domain label size ehckers should only be applied for fully qualified domain names.
+    // FQDNs have label hierarchy (meaning at least one dot).
+    if(host.value().find_first_of(".") != std::string_view::npos) {
```
Comment (lemire) [inline src/url.cpp:480]: Possibly `host.value().find_first_of(".")` to `host.value().find_first_of('.')`. I don't know that it actually makes a difference... but I expect that '.' might be faster (???). I also tend to prefer "find('.')" because it is shorter... though it might not matter.

File: src/url.cpp [hunk 2]
```diff
@@ -469,6 +474,13 @@ namespace ada {
     if(checkers::is_ipv4(host.value())) {
       return parse_ipv4(host.value());
     }
+
+    // The domain label size ehckers should only be applied for fully qualified domain names.
```
Comment (lemire) [inline src/url.cpp:478]:
```suggestion
    // The domain label size checkers should only be applied for fully qualified domain names.
```

File: src/checkers.cpp [hunk 3]
```diff
@@ -51,5 +51,36 @@ namespace ada::checkers {
...
+        // Find the next dot in the domain
+        const char* dot = std::find(start, end, '.');
+
+        // Calculate the size of the current label
+        auto size = dot - start;
+        if (size > 63 || size == 0) {
+            return false;
+        }
+
+        ++dot_count;
```
Comment (lemire) [inline src/checkers.cpp:74]: Technically, this does not count dots, but labels. (Not a request for change.)

File: src/url.cpp (hunk 1 above)
Comment (lemire) [reply src/url.cpp:480]: (This is not a request for change... your `host.value().find_first_of(".")` is probably? fine.)

File: src/checkers.cpp (hunk 3 above)
Comment (miguelteixeiraa) [reply src/checkers.cpp:74]: You are right :sweat_smile:

File: src/url.cpp (hunk 1 above)
Comment (lemire) [inline src/url.cpp:480]: Why not do `checkers::check_domain(host.value());` directly? The if clause is almost always going to be true anyhow, so it saves us nothing, and you are inlining check_domain. In check_domain, you could easily acomodate the case where there is no dot: just unroll your loop once: process the first search for a '.' separately. In this way, you will avoid doing the work twice.

**Comment by lemire** [comment]
I love this PR. :heart:

**Comment by anonrig** [comment]
I opened an issue on this to Node.

@miguelteixeiraa Can you add test example for this issue, and remove the buffer overflow in to_ascii method, since we don't want URLs with more than 255 characters.

File: src/checkers.cpp [hunk 4]
```diff
@@ -51,5 +51,36 @@ namespace ada::checkers {
...
     return accumulator;
   }
+
+  ada_really_inline constexpr bool check_domain(std::string_view input) noexcept {
+    if(input.size() > 255) {
+      return false;
+    }
+
+    const char* start = input.data();
+    const char* end = start + input.size();
```
Comment (anonrig) [inline src/checkers.cpp:61]: Why don't we use `std::string_view::iterator`?

File: src/checkers.cpp [hunk 5]
```diff
@@ -51,5 +51,36 @@ namespace ada::checkers {
...
+
+    int dot_count = 0;
+    while (start < end) {
+        // Find the next dot in the domain
+        const char* dot = std::find(start, end, '.');
```
Comment (anonrig) [inline src/checkers.cpp:66]: I believe `size_t dot_location = input.find('.', left_padding);` would reduce the number of instructions.

**Comment by miguelteixeiraa** [comment]
> 

So, currently these validations are already being done by the ICU. Is that correct (at least in theory)?
So we just need to add tests to see if they pass

**Comment by miguelteixeiraa** [comment]
> I opened an issue on this to Node.
> 
> @miguelteixeiraa Can you add test example for this issue, and remove the buffer overflow in to_ascii method, since we don't want URLs with more than 255 characters.

Yeah I'll add test examples!

File: include/ada/checkers.h [hunk 6]
```diff
@@ -93,6 +93,13 @@ namespace ada::checkers {
    */
   ada_really_inline constexpr uint8_t path_signature(std::string_view input) noexcept;
 
+
+  /** 
+  * A domain cannot be more than 255 octets and have more than 127 labels, each label must have
+  * a maximum of 63 octets.
```
Comment (anonrig) [inline include/ada/checkers.h:99]: Can you add a reference to the specification using `@see`. You might give a context why this is the case and mention:

1. URL spec calls ToASCII with be_strict true
2. ToASCII passes `VerifyDnsLength` if `be_strict` is true.
3. `VerifyDnsLength` checks 253 and 63 limitations (referencing: https://www.unicode.org/reports/tr46/#ToASCII)

Comment (anonrig) [reply include/ada/checkers.h:99]: Let's be as open as possible in here, to avoid communication issues.

File: src/checkers.cpp [hunk 7]
```diff
@@ -51,5 +51,36 @@ namespace ada::checkers {
     }
     return accumulator;
   }
+
+  ada_really_inline constexpr bool check_domain(std::string_view input) noexcept {
+    if(input.size() > 255) {
```
Comment (anonrig) [inline src/checkers.cpp:56]: Why is it 255?

```
The length of the domain name, excluding the root label and its dot, is from 1 to 253.
```

File: src/checkers.cpp [hunk 8]
```diff
@@ -51,5 +51,36 @@ namespace ada::checkers {
...
+    const char* end = start + input.size();
+
+    int dot_count = 0;
+    while (start < end) {
+        // Find the next dot in the domain
+        const char* dot = std::find(start, end, '.');
+
+        // Calculate the size of the current label
+        auto size = dot - start;
+        if (size > 63 || size == 0) {
```
Comment (anonrig) [inline src/checkers.cpp:70]: Can you add a comment here or to the header file declaration that `When VerifyDnsLength is true, the empty root label is disallowed.`

File: src/url.cpp [hunk 9]
```diff
@@ -458,6 +458,11 @@ namespace ada {
         if (checkers::is_ipv4(host.value())) {
           return parse_ipv4(host.value());
         }
+
+        if(host.value().find_first_of(".") != std::string_view::npos) {
+          return checkers::check_domain(host.value());
+        }
+        
         return true;
```
Comment (anonrig) [inline src/url.cpp:466]: You can just:

`return checkers::check_domain(host.value())`

File: src/url.cpp (hunk 1 above)
Comment (anonrig) [reply src/url.cpp:480]: I agree with @lemire: We can just return checkers::check_ domain and remove the branch.

File: include/ada/checkers.h [hunk 10]
```diff
@@ -93,6 +93,13 @@ namespace ada::checkers {
...
+  */
+  ada_really_inline constexpr bool check_domain(std::string_view input) noexcept;
```
Comment (anonrig) [inline include/ada/checkers.h:101]: Can we use a more spec specific naming for this function like: `verify_dns_length`

**Comment by anonrig** [comment]
We are also calling to_ascii with `be_strict` false, which is wrong.

**Comment by lemire** [comment]
> So, currently these validations are already being done by the ICU. Is that correct (at least in theory)?
So we just need to add tests to see if they pass

Most times, we bypass ICU so your PR is needed. 

Furthermore, though ICU can run these tests, we seem to be ignoring its errors deliberately.

**Comment by lemire** [comment]
To be clear, @miguelteixeiraa : your PR is needed. It is not work in vain.

**Comment by miguelteixeiraa** [comment]
> I opened an issue on this to Node.
> 
> @miguelteixeiraa Can you add test example for this issue, and remove the buffer overflow in to_ascii method, since we don't want URLs with more than 255 characters.

I think that these limits are only for the domain names (like the www.example.com in the www.example.com/path). I believe that there is also a limit for the URL (I know it is more than 2000 :sweat_smile:) . So maybe remove the buffer overflow + add a check for this "max string to be evaluated" aka "URL" ?

File: include/ada/checkers.h (hunk 10 above)
Comment (miguelteixeiraa) [reply include/ada/checkers.h:101]: Sure thing!

**Comment by anonrig** [comment]
> I believe that there is also a limit for the URL (I know it is more than 2000 😅)

I couldn't find a reference in the URL spec to `2048` limit. What's your source?

File: src/checkers.cpp (hunk 7 above)
Comment (miguelteixeiraa) [reply src/checkers.cpp:56]: Yeah, that was my doubt. I was considering the 255 octets (bytes) as 255 chars. But I saw somewhere something about the last 2 being like "\0" or whatever other thing that means they arent "working" positions. I'll move to 253

File: include/ada/checkers.h (hunk 6 above)
Comment (miguelteixeiraa) [reply include/ada/checkers.h:99]: Thanks for the link!
I was considering:

```
labels          63 octets or less

names           255 octets or less
```
in  https://www.rfc-editor.org/rfc/rfc1035

**Comment by miguelteixeiraa** [comment]
> > I believe that there is also a limit for the URL (I know it is more than 2000 😅)
> 
> I couldn't find a reference in the URL spec to `2048` limit. What's your source?

My bad. I didn't read this in any spec and just checked that this 2xxx thing is a "browsers specific" thing. Also seems that there is not really a spec for the size of an URL.

File: src/checkers.cpp (hunk 5 above)
Comment (lemire) [reply src/checkers.cpp:66]: We have the instrumentation to count the instructions, so we can explore the design space as needed.

**Comment by lemire** [comment]
Related: https://github.com/dotnet/runtime/issues/33746

File: src/checkers.cpp (hunk 7 above)
Comment (lemire) [reply src/checkers.cpp:56]: It is wrong.

Look at UCI:

https://github.com/unicode-org/icu/blob/bb0e745e25c99cc57055caf45c81b95ef63b25d4/icu4c/source/common/uts46.cpp#LL240C44-L240C76

If there is a dot in final position, then it we allow 254.

**Comment by lemire** [comment]
> Also seems that there is not really a spec for the size of an URL.

I think that there is none but we could enforce a limit if we wanted to, we would just need to make it a public-facing parameter. I personally think that 4MB is too much.

**Comment by miguelteixeiraa** [comment]
Sorry I started a review by accident :joy:

**Comment by lemire** [comment]
@miguelteixeiraa I think what we want is a strict limit of 254 bytes if the domain name ends with '.', but only 253 otherwise (that's for FQ).

**Comment by lemire** [comment]
@miguelteixeiraa I do think that `example.com.` ought to be valid.

>So for instance if we look at the true DNS representation of webmasters.stackexchange.com it is actually webmasters.stackexchange.com. with the trailing dot. This empty label that comes after the TLD is always reserved for the root, and cannot be used anywhere else in the domain name.

source: https://webmasters.stackexchange.com/questions/119097/are-multiple-consecutive-dots-allowed-in-a-domain-name


I think that how it works is that we have a budget of 255 bytes. We always have a final dot. And we use one byte to store the length of the domain.

**Review by anonrig — DISMISSED** [review]
Really, really good. I'm impressed!

**Comment by lemire** [comment]
Merging to a side branch.
